test/e2e/finalizers_test.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	capi_e2e "sigs.k8s.io/cluster-api/test/e2e"
+	"sigs.k8s.io/cluster-api/test/framework"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
+)
+
+var _ = Describe("Check finalizers resiliency with DeploymentZone and ClusterIdentity", func() {
+	// Before running the test create the secret used by the VSphereClusterIdentity to connect to the vCenter.
+	BeforeEach(func() {
+		createVsphereIdentitySecret(ctx, bootstrapClusterProxy)
+	})
+
+	capi_e2e.QuickStartSpec(ctx, func() capi_e2e.QuickStartSpecInput {
+		return capi_e2e.QuickStartSpecInput{
+			E2EConfig:             e2eConfig,
+			ClusterctlConfigPath:  clusterctlConfigPath,
+			BootstrapClusterProxy: bootstrapClusterProxy,
+			ArtifactFolder:        artifactFolder,
+			SkipCleanup:           skipCleanup,
+			Flavor:                ptr.To("finalizers"),
+			PostMachinesProvisioned: func(proxy framework.ClusterProxy, namespace, clusterName string) {
+				// Set up a periodic patch to ensure the DeploymentZone is reconciled.
+				reconcileDeploymentZonePeriodically(ctx, proxy.GetClient())
+
+				// This check ensures that finalizers are resilient - i.e. correctly re-reconciled, when removed.
+				framework.ValidateFinalizersResilience(ctx, proxy, namespace, clusterName,
+					framework.CoreFinalizersAssertion,
+					framework.KubeadmControlPlaneFinalizersAssertion,
+					framework.ExpFinalizersAssertion,
+					VSphereFinalizers,
+				)
```

Comment:
Looks like the ownerreference test already has what we need (from object perspective).

Could we combine these two tests maybe to combine ownerreferences and finalizers in one test?
If I got it right its only about having all the objects and running `framework.ValidateFinalizersResilience` as `PostMachinesProvisioned`.
This would reduce the time needed for CI because bringing up cluster takes time and costs resources :-)

Or do you see points why we should not combine them?

Reply:
Yes combining these tests will surely make it more efficient, should also work (looks good locally). 

I have purposely made a second commit to see the notable differences, before and after test combination and also renamed the flavor `ownerreferences` -> `ownerrefs-finalizers` which IMO reflects the purpose of the combined test. https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/2626/commits/42a1cff67f25dcf96398ad112006c2f6e1aed955

Once these changes are reviewed I will squash my commits.